File: Freedom/微信/我/ExpressionViewController/WXExpressionPublicViewController.m
//  TLExpressionPublicViewController.m
//  Freedom
// Created by Super
#import "WXExpressionPublicViewController.h"
#import "WXExpressionSearchViewController.h"
#import "WXSearchController.h"
#import "WXExpressionDetailViewController.h"
#define         EDGE                20.0
#define         SPACE_CELL          EDGE
#define         WIDTH_CELL          ((APPW - EDGE * 2 - SPACE_CELL * 2.0) / 3.0)
#define         HEIGHT_CELL         (WIDTH_CELL + 20)
#import "WXExpressionHelper.h"
#import <UIKit/UIKit.h>
#import "TLEmojiBaseCell.h"
@interface WXExpressionPublicCell : UICollectionViewCell
@property (nonatomic, strong) TLEmojiGroup *group;
@end
@interface WXExpressionPublicCell ()
@property (nonatomic, strong) UIImageView *imageView;
@property (nonatomic, strong) UILabel *titleLabel;
@end
@implementation WXExpressionPublicCell
- (id)initWithFrame:(CGRect)frame{
    if (self = [super initWithFrame:frame]) {
        [self.contentView addSubview:self.imageView];
        [self.contentView addSubview:self.titleLabel];
        [self p_addMasonry];
    }
    return self;
}
- (void)setGroup:(TLEmojiGroup *)group{
    _group = group;
    [self.titleLabel setText:group.groupName];
    UIImage *image = [UIImage imageNamed:group.groupIconPath];
    if (image) {
        [self.imageView setImage:image];
    }else{
        [self.imageView sd_setImageWithURL:[NSURL URLWithString:group.groupIconURL] placeholderImage:[FreedomTools imageWithColor:[UIColor lightGrayColor]]];
    }
}
#pragma mark - Private Methods
- (void)p_addMasonry{
    [self.imageView mas_makeConstraints:^(MASConstraintMaker *make) {
        make.left.and.top.and.right.mas_equalTo(self.contentView);
        make.height.mas_equalTo(self.imageView.mas_width);
    }];
    [self.titleLabel mas_makeConstraints:^(MASConstraintMaker *make) {
        make.centerX.mas_equalTo(self.contentView);
        make.top.mas_equalTo(self.imageView.mas_bottom).mas_offset(7.0f);
        make.width.mas_lessThanOrEqualTo(self.contentView);
    }];
}
#pragma mark - Getter
- (UIImageView *)imageView{
    if (_imageView == nil) {
        _imageView = [[UIImageView alloc] init];
        [_imageView setUserInteractionEnabled:NO];
        [_imageView.layer setMasksToBounds:YES];
        [_imageView.layer setCornerRadius:5.0f];
        [_imageView.layer setBorderWidth:1];
        [_imageView.layer setBorderColor:[UIColor grayColor].CGColor];
    }
    return _imageView;
}
- (UILabel *)titleLabel{
    if (_titleLabel == nil) {
        _titleLabel = [[UILabel alloc] init];
        [_titleLabel setTextAlignment:NSTextAlignmentCenter];
        [_titleLabel setFont:[UIFont systemFontOfSize:12.0f]];
    }
    return _titleLabel;
}
@end
@interface WXExpressionPublicViewController () <UISearchBarDelegate>
@property (nonatomic, strong) WXSearchController *searchController;
@property (nonatomic, strong) WXExpressionSearchViewController *searchVC;
@property (nonatomic,strong) UICollectionView *collectionView;
@end
@implementation WXExpressionPublicViewController
- (void)viewDidLoad{
    [super viewDidLoad];
    [self setAutomaticallyAdjustsScrollViewInsets:NO];
    
    UICollectionViewFlowLayout *layout = [[UICollectionViewFlowLayout alloc] init];
    CGRect rect = CGRectMake(0, TopHeight + 20, APPW, APPH - TopHeight - 20);
    self.collectionView = [[UICollectionView alloc] initWithFrame:rect collectionViewLayout:layout];
    [self.collectionView setBackgroundColor:[UIColor whiteColor]];
    [self.collectionView setDataSource:self];
    [self.collectionView setDelegate:self];
    [self.collectionView setAlwaysBounceVertical:YES];
    
    [self.view addSubview:self.collectionView];
    [self.collectionView setBackgroundColor:[UIColor whiteColor]];
//    [self.collectionView setTableHeaderView:self.searchController.searchBar];
    
    MJRefreshAutoNormalFooter *footer = [MJRefreshAutoNormalFooter footerWithRefreshingTarget:self refreshingAction:@selector(loadMoreData)];
    [footer setTitle:@"正在加载..." forState:MJRefreshStateRefreshing];
    [footer setTitle:@"" forState:MJRefreshStateNoMoreData];
    [self.collectionView setMj_footer:footer];
    
    [self registerCellForCollectionView:self.collectionView];
    [self loadDataWithLoadingView:YES];
}
- (void)viewDidDisappear:(BOOL)animated{
    [super viewDidDisappear:animated];
    [SVProgressHUD dismiss];
}
#pragma mark - Delegate
//MARK: UISearchBarDelegate
- (void)searchBarTextDidBeginEditing:(UISearchBar *)searchBar{
}
- (void)searchBarTextDidEndEditing:(UISearchBar *)searchBar{
    
}
#pragma mark - Getter
- (WXSearchController *)searchController{
    if (_searchController == nil) {
        _searchController = [[WXSearchController alloc] initWithSearchResultsController:self.searchVC];
        [_searchController setSearchResultsUpdater:self.searchVC];
        [_searchController.searchBar setPlaceholder:@"搜索表情"];
        [_searchController.searchBar setDelegate:self];
    }
    return _searchController;
}
- (WXExpressionSearchViewController *)searchVC{
    if (_searchVC == nil) {
        _searchVC = [[WXExpressionSearchViewController alloc] init];
    }
    return _searchVC;
}
- (WXExpressionHelper *)proxy{
    if (_proxy == nil) {
        _proxy = [WXExpressionHelper sharedHelper];
    }
    return _proxy;
}
- (void)registerCellForCollectionView:(UICollectionView *)collectionView{
    [collectionView registerClass:[WXExpressionPublicCell class] forCellWithReuseIdentifier:@"TLExpressionPublicCell"];
    [collectionView registerClass:[UICollectionViewCell class] forCellWithReuseIdentifier:@"EmptyCell"];
}
#pragma mark - Delegate
- (NSInteger)collectionView:(UICollectionView *)collectionView numberOfItemsInSection:(NSInteger)section{
    return self.data.count == 0 ? 1 : self.data.count;
}
- (UICollectionViewCell *)collectionView:(UICollectionView *)collectionView cellForItemAtIndexPath:(NSIndexPath *)indexPath{
    if (indexPath.row < self.data.count) {
        WXExpressionPublicCell *cell = [collectionView dequeueReusableCellWithReuseIdentifier:@"TLExpressionPublicCell" forIndexPath:indexPath];
        TLEmojiGroup *emojiGroup = [self.data objectAtIndex:indexPath.row];
        [cell setGroup:emojiGroup];
        return cell;
    }
    return [collectionView dequeueReusableCellWithReuseIdentifier:@"EmptyCell" forIndexPath:indexPath];
}
- (void)collectionView:(UICollectionView *)collectionView didSelectItemAtIndexPath:(NSIndexPath *)indexPath{
    if (indexPath.row < self.data.count) {
        TLEmojiGroup *group = [self.data objectAtIndex:indexPath.row];
        WXExpressionDetailViewController *detailVC = [[WXExpressionDetailViewController alloc] init];
        [detailVC setGroup:group];
        [self.parentViewController setHidesBottomBarWhenPushed:YES];
        [self.parentViewController.navigationController pushViewController:detailVC animated:YES];
    }
    [collectionView deselectItemAtIndexPath:indexPath animated:NO];
}
- (CGSize)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout *)collectionViewLayout sizeForItemAtIndexPath:(NSIndexPath *)indexPath{
    if (indexPath.row < self.data.count) {
        return CGSizeMake(WIDTH_CELL, HEIGHT_CELL);
    }
    return collectionView.bounds.size;
}
- (CGFloat)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout *)collectionViewLayout minimumLineSpacingForSectionAtIndex:(NSInteger)section{
    return self.data.count == 0 ? 0 : SPACE_CELL;
}
- (CGFloat)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout *)collectionViewLayout minimumInteritemSpacingForSectionAtIndex:(NSInteger)section{
    return self.data.count == 0 ? 0 : SPACE_CELL;
}
- (UIEdgeInsets)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout *)collectionViewLayout insetForSectionAtIndex:(NSInteger)section{
    return self.data.count == 0 ? UIEdgeInsetsMake(0, 0, 0, 0) : UIEdgeInsetsMake(EDGE, EDGE, EDGE, EDGE);
}
- (void)loadDataWithLoadingView:(BOOL)showLoadingView{
    if (showLoadingView) {
        [SVProgressHUD show];
    }
    kPageIndex = 1;
    __weak typeof(self) weakSelf = self;
    [self.proxy requestExpressionPublicListByPageIndex:kPageIndex success:^(id data) {
        [SVProgressHUD dismiss];
        kPageIndex ++;
        weakSelf.data = [[NSMutableArray alloc] init];
        for (TLEmojiGroup *group in data) {     // 优先使用本地表情
            TLEmojiGroup *localEmojiGroup = [[WXExpressionHelper sharedHelper] emojiGroupByID:group.groupID];
            if (localEmojiGroup) {
                [self.data addObject:localEmojiGroup];
            }else{
                [self.data addObject:group];
            }
        }
        [weakSelf.collectionView reloadData];
    } failure:^(NSString *error) {
        [SVProgressHUD dismiss];
    }];
}
- (void)loadMoreData{
    __weak typeof(self) weakSelf = self;
    [self.proxy requestExpressionPublicListByPageIndex:kPageIndex success:^(NSMutableArray *data) {
        [SVProgressHUD dismiss];
        if (data.count == 0) {
            [self.collectionView.mj_footer endRefreshingWithNoMoreData];
        }else{
            [self.collectionView.mj_footer endRefreshing];
            kPageIndex ++;
            for (TLEmojiGroup *group in data) {     // 优先使用本地表情
                TLEmojiGroup *localEmojiGroup = [[WXExpressionHelper sharedHelper] emojiGroupByID:group.groupID];
                if (localEmojiGroup) {
                    [self.data addObject:localEmojiGroup];
                }
                else {
                    [self.data addObject:group];
                }
            }
            [weakSelf.collectionView reloadData];
        }
    } failure:^(NSString *error) {
        [self.collectionView.mj_footer endRefreshingWithNoMoreData];
        [SVProgressHUD dismiss];
    }];
}
@end

File: Freedom/微信/我/ExpressionViewController/WXExpressionPublicViewController.swift
//
//  WXExpressionPublicViewController.swift
//  Freedom
import SnapKit
import MJRefresh
import XCarryOn
import XExtension
import Foundation
class WXExpressionPublicCell: UICollectionViewCell {
    var group: TLEmojiGroup = TLEmojiGroup() {
        didSet {
            titleLabel.text = group.groupName
            let image = group.groupIconPath.image
            if image != nil {
                imageView.image = image
            } else {
                imageView.sd_setImage(with: URL(string: group.groupIconURL), placeholderImage: UIImage.imageWithColor(UIColor.light))
            }
        }
    }
    lazy var imageView: UIImageView = {
        let imageView = UIImageView()
        imageView.isUserInteractionEnabled = false
        imageView.layer.masksToBounds = true
        imageView.layer.cornerRadius = 5.0
        imageView.layer.borderWidth = 1
        imageView.layer.borderColor = UIColor.grayx.cgColor
        return imageView
    }()
    lazy var titleLabel: UILabel = {
        let titleLabel = UILabel()
        titleLabel.textAlignment = .center
        titleLabel.font = UIFont.small
        return titleLabel
    }()
    override init(frame: CGRect) {
        super.init(frame: frame)
        contentView.addSubview(imageView)
        contentView.addSubview(titleLabel)
        imageView.snp.makeConstraints { (make) in
            make.left.top.right.equalTo(self.contentView)
            make.height.equalTo(self.imageView.snp.width)
        }
        titleLabel.snp.makeConstraints { (make) in
            make.centerX.equalTo(self.contentView)
            make.top.equalTo(self.imageView.snp.bottom).offset(7)
            make.width.lessThanOrEqualTo(self.contentView)
        }
    }
    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
    }
}
class WXExpressionPublicViewController: WXBaseViewController, UICollectionViewDelegate, UICollectionViewDataSource, UICollectionViewDelegateFlowLayout, UISearchBarDelegate {
    var collectionView: BaseCollectionView!
    var kPageIndex: Int = 0
    var data: [TLEmojiGroup] = []
    var proxy = WXExpressionHelper.shared
    lazy var searchController: WXSearchController = {
        let searchController = WXSearchController(searchResultsController: searchVC)
        searchController.searchResultsUpdater = searchVC
        searchController.searchBar.placeholder = "搜索表情"
        searchController.searchBar.delegate = self
        return searchController
    }()
    var searchVC = WXExpressionSearchViewController()
    override func viewDidLoad() {
        super.viewDidLoad()
        automaticallyAdjustsScrollViewInsets = false
        let layout = UICollectionViewFlowLayout()
        let rect = CGRect(x: 0, y: TopHeight + 20, width: APPW, height: APPH - TopHeight - 20)
        collectionView = BaseCollectionView(frame: rect, collectionViewLayout: layout)
        collectionView.backgroundColor = UIColor.whitex
        collectionView.dataSource = self
        collectionView.delegate = self
        collectionView.alwaysBounceVertical = true
        view.addSubview(collectionView)
        collectionView.backgroundColor = UIColor.whitex
        //    [self.collectionView setTableHeaderView:self.searchController.searchBar];
        let footer = MJRefreshAutoNormalFooter(refreshingTarget: self, refreshingAction: #selector(self.loadMoreData))
        footer?.setTitle("正在加载...", for: .refreshing)
        footer?.setTitle("", for: .noMoreData)
        collectionView.mj_footer = (footer)
        collectionView.register(WXExpressionPublicCell.self, forCellWithReuseIdentifier: WXExpressionPublicCell.identifier)
        collectionView.register(UICollectionViewCell.self, forCellWithReuseIdentifier: UICollectionViewCell.identifier)
        loadData(withLoadingView: true)
    }
    override func viewDidDisappear(_ animated: Bool) {
        super.viewDidDisappear(animated)
        XHud.hide()
    }
    func searchBarTextDidBeginEditing(_ searchBar: UISearchBar) {
    }
    func searchBarTextDidEndEditing(_ searchBar: UISearchBar) {
    }
    func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        return data.count == 0 ? 1 : data.count
    }
    func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
        if indexPath.row < data.count {
            let cell = collectionView.dequeueCell(WXExpressionPublicCell.self, for: indexPath)
            let emojiGroup = data[indexPath.row] as TLEmojiGroup
            cell.group = emojiGroup
            return cell
        }
        return collectionView.dequeueCell(UICollectionViewCell.self, for: indexPath)
    }

    func collectionView(_ collectionView: UICollectionView, didSelectItemAt indexPath: IndexPath) {
        if indexPath.row < data.count {
            let group = data[indexPath.row] as TLEmojiGroup
            let detailVC = WXExpressionDetailViewController()
            detailVC.group = group
            parent?.navigationController?.pushViewController(detailVC, animated: true)
        }
        collectionView.deselectItem(at: indexPath, animated: false)
    }
    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, sizeForItemAt indexPath: IndexPath) -> CGSize {
        if indexPath.row < data.count {
            return CGSize(width: (APPW - 20 * 2 - 20 * 2.0) / 3.0, height: ((APPW - 20 * 2 - 20 * 2.0) / 3.0) + 20)
        }
        return collectionView.bounds.size
    }

    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, minimumLineSpacingForSectionAt section: Int) -> CGFloat {
        return data.count == 0 ? 0 : 20
    }

    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, minimumInteritemSpacingForSectionAt section: Int) -> CGFloat {
        return data.count == 0 ? 0 : 20
    }
    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, insetForSectionAt section: Int) -> UIEdgeInsets {
        return data.count == 0 ? UIEdgeInsets(top: 0, left: 0, bottom: 0, right: 0) : UIEdgeInsets(top: 20, left: 20, bottom: 20, right: 20)
    }

    func loadData(withLoadingView showLoadingView: Bool) {
        if showLoadingView {
            XHud.show()
        }
        kPageIndex = 1
        XNetKit.kit.requestExpressionPublicList(byPageIndex: kPageIndex, success: { data in
            XHud.hide()
            self.kPageIndex += 1
            for group: TLEmojiGroup in data {
                // 优先使用本地表情
                let localEmojiGroup = WXExpressionHelper.shared.emojiGroup(byID: group.groupID)
                if let localEmojiGroup = localEmojiGroup {
                    self.data.append(localEmojiGroup)
                } else {
                    self.data.append(group)
                }
            }
            self.collectionView.reloadData()
        }, failure: { error in
            XHud.hide()
        })
    }

    func loadMoreData() {
        XNetKit.kit.requestExpressionPublicList(byPageIndex: kPageIndex, success: { data in
            XHud.hide()
            if data.count == 0 {
                self.collectionView.mj_footer.endRefreshingWithNoMoreData()
            } else {
                self.collectionView.mj_footer.endRefreshing()
                self.kPageIndex += 1
                for group: TLEmojiGroup in data {
                    // 优先使用本地表情
                    let localEmojiGroup = WXExpressionHelper.shared.emojiGroup(byID: group.groupID)
                    if let localEmojiGroup = localEmojiGroup {
                        self.data.append(localEmojiGroup)
                    } else {
                        self.data.append(group)
                    }
                }
                self.collectionView.reloadData()
            }
        }, failure: { error in
            self.collectionView.mj_footer.endRefreshingWithNoMoreData()
            XHud.hide()
        })
    }
}
